### PR TITLE
Logistic regression

### PR DESCRIPTION
**Pull request by PeterDamrosch**

### CODE REVIEW

**Comment by cavaunpeu** [comment]
:eyes: @cavaunpeu

File: logistic_regression.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	return prob
+
+# Example use of logistic_function
+print logistic_function(log_coeff,720,10000)
+
+# Visualize the model output with a range of FICO scores at $10,000 requested
+x = range(600,800)
+y = [logistic_function(log_coeff, i, 10000) for i in x]
+plt.plot(x,y)
+plt.show()
```
Comment (cavaunpeu) [inline logistic_regression.py:45]: all files in every language should end with a blank line. this is what the red circle slash thing indicates.

**Comment by PeterDamrosch** [comment]
I think I've made a breakthrough in logistic regression 👀  cavaunpeu

File: logistic_regression.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+log_coeff = result.params
+
+# Find P(interest rate < 12%) given logistic model coefficients, FICo score, and amount requested
+def logistic_function(coeff, score, amount):
+	a = coeff['Intercept'] + coeff['FICO.Score'] * score + coeff['Amount.Requested'] * amount
+	prob = 1/(1 + math.e ** (-1 * a))
+	return prob
+
+# Example use of logistic_function
+print logistic_function(log_coeff,720,10000)
```
Comment (cavaunpeu) [inline logistic_regression.py:39]: let's use python 3! `print()` would be the syntax.

Comment (PeterDamrosch) [inline logistic_regression.py:39]: isn't python 3 for squares?

File: logistic_regression.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import pandas as pd
+import statsmodels.api as sm
+import matplotlib.pyplot as plt
+
+# Load data on loans
+loans = pd.read_csv("loans_clean.csv")
+
+# Create boolean column for interest rate < 12%
+# I'm used to defining functions out like this, but there's probably something quicker for pandas
+def less_than_12(rate):
```
Comment (cavaunpeu) [inline logistic_regression.py:11]: i don't immediately know what `less_than_12`. let's try for more descriptive function names.

File: logistic_regression.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	if rate < 0.12:
+		return 1
+	else:
+		return 0
+
+loans['IR_TF'] = loans['Interest.Rate'].map(less_than_12)
```
Comment (cavaunpeu) [inline logistic_regression.py:17]: we can do this MUCH more easily.

`loans['IR_TF'] = loans['Interest.Rate'] < .12`

File: logistic_regression.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+logit = sm.Logit(loans['IR_TF'], loans[ind_vars])
+result = logit.fit()
+
+# Create coefficient object for our model
+log_coeff = result.params
+
+# Find P(interest rate < 12%) given logistic model coefficients, FICo score, and amount requested
+def logistic_function(coeff, score, amount):
+	a = coeff['Intercept'] + coeff['FICO.Score'] * score + coeff['Amount.Requested'] * amount
+	prob = 1/(1 + math.e ** (-1 * a))
```
Comment (cavaunpeu) [inline logistic_regression.py:35]: i'd advise using numpy. everyone likes numpy. in numpy, we can rewrite this as `1/(1 + np.exp(-a))`

File: homemade_regression.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+def linear_regression(x, y):
+	# Define constants so don't have to recalculate
+	mean_y, mean_x = y.mean(), x.mean()
+	std_y, std_x = y.std(), x.std()
+
+	# Pearson's R calculations using Z-scores
+	correlation_sum = 0
+	for i in range(len(y)):
+		z_y = (y[i] - mean_y) / std_y
+		z_x = (x[i] - mean_x) / std_x
```
Comment (cavaunpeu) [inline homemade_regression.py:15]: we can do this with pandas Series and a `.map` or `.transform` method, no? that way, we can take advantage of some of numpy's vectorized calculations under the hood, instead of iterating one-by-one through a generator of numbers as you have on line 13. finally, you could multiply the series together and then call `.sum()`.

File: cricketsData.csv [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+15.5,75.19999695
+14.69999981,69.69999695
+17.10000038,82
+15.39999962,69.40000153
+16.20000076,83.30000305
+15,79.59999847
+17.20000076,82.59999847
+16,80.59999847
+17,83.5
+14.39999962,76.30000305
```
Comment (cavaunpeu) [inline cricketsData.csv:16]: newline.

File: logistic_regression.py (hunk 2 above)
Comment (cavaunpeu) [inline logistic_regression.py:39]: not in the slightest. in fact, Python 2 will be fully deprecated in the coming years. Python 3 1000%.

**Comment by cavaunpeu** [comment]
homemade regression looks great. for this type of exercise, I think having an [integration test](https://www.fullstackpython.com/integration-testing.html) is extremely important; for example, comparing the output of scikit-learn's [LinearRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html) object and your own.

if unfamiliar with integration and/or unit testing, let's cover it on Tuesday!
